recipes-extended/xen/files/0001-convert-block-tap-for-shared.patch

```diff
@@ -0,0 +1,551 @@
++date >&2
++PS4="+ $$: "
++set -x
++
+ remove_label()
+ {
+     echo $1 | sed "s/^\("$2"\)//"
+@@ -81,23 +86,135 @@ find_device()
+     done
+ 
+     if [ -z "$pid" ] || [ -z "$minor" ]; then
+-        fatal "cannot find required parameters"
++        return 1
++    fi
++
++    return 0
++}
++
++##
++# count_sharing device mode
++#
++# Check whether the device requested is already in use.  To use the device in
++# read-only mode, it may be in use in read-only mode, but may not be in use in
++# read-write anywhere at all.  To use the device in read-write mode, it must
++# not be in use anywhere at all.
++#
++# Prints one of
++#
++#    'local $d': the device ($d) may not be used because it is mounted in the
++#                current (i.e. the privileged domain) in a way incompatible
++#                with the requested mode;
++#    'guest $d': the device may not be used because it is already mounted
++#                through device $d by a guest in a way incompatible with the
++#                requested mode; or
++#    'ok':       the device may be used.
++#
++count_using()
++{
++  local dev="$1"
++  local mode="$2"
++  local devmm=","
++
++  devmm=${devmm}$(device_major_minor "$dev")","
++
++  local file
++
++  if [ "$mode" = 'w' ]
++  then
++    toskip="^$"
++  else
++    toskip="^[^ ]* [^ ]* [^ ]* ro[, ]"
++  fi
++
++  for file in $(cat /proc/mounts | grep -v "$toskip" | cut -f 1 -d ' ')
++  do
++    if [ -e "$file" ]
++    then
++      local d=$(device_major_minor "$file")
++
++      # checking for $d in $devmm is best through the [[...]] bashism
++      if [[ "$devmm" == *",$d,"* ]]
++      then
++        echo "local $d"
++        return
++      fi
+     fi
++  done
++
++  local i=0
++  local base_path="$XENBUS_BASE_PATH/$XENBUS_TYPE"
++  for dom in $(xenstore-list "$base_path")
++  do
++    for dev in $(xenstore-list "$base_path/$dom")
++    do
++      d=$(xenstore_read_default "$base_path/$dom/$dev/physical-device" "")
++
++      # checking for $d in $devmm is best through the [[...]] bashism
++      if [ -n "$d" ] && [[ "$devmm" == *",$d,"* ]]
++      then
++          if ! same_vm $dom
++          then
++              i=$(( i + 1 ))
++          fi
++      fi
++    done
++  done
++
++  echo "$i"
++}
++
++is_shared()
++{
++    local devpath="/dev/xen/blktap-2/tapdev$minor"
++
++    [ $( count_using $devpath $mode ) -gt 1 ]
+ }
+ 
+ # Attaches the device and writes xenstore backend entries to connect
+ # the device
+ add()
+ {
+-    dev=$(tap-ctl create -a $target)
++    local tap_opts
++
++    if [ "$mode" = "r" ]; then
++        tap_opts="-R"
++    fi
++
++    claim_lock "block"
++    if find_device; then
++        dev="/dev/xen/blktap-2/tapdev$minor"
++        check_device_sharing "$dev" "$mode"
++    else
++        TAPDISK3_CRYPTO_KEYDIR=$( db-read /xenmgr/platform-crypto-key-dirs )
++        : ${TAPDISK3_CRYPTO_KEYDIR:=/config/platform-crypto-keys}
++        USER_KEYDIRS=$( xec-vm -n uivm get crypto-key-dirs )
```

Comment:
This was what we were avoiding by calling tapEnv in xenmgr, is this a separate case where the block script is being called outside of xenmgr?

Reply:
It gets removed here:
commit c8547eed43cd1764b87701b1b47dfe8a6fb327c5
Author: Jason Andryuk <jandryuk@gmail.com>
Date:   Fri Nov 19 12:02:39 2021 -0500

    xen: convert-block-tap-for-shared.patch

    count_sharing needs some more updating.

    Signed-off-by: Jason Andryuk <jandryuk@gmail.com>

The commit message doesn't mention it :/

Reply:
Gotcha.